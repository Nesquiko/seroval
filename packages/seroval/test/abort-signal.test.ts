import { describe, expect, it } from 'vitest';
import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from '../src';

const delayedAbortSignal = () => {
  const controller = new AbortController();
  setTimeout(() => {
    controller.abort('aborted!');
  }, 100);
  return controller.signal;
};

const SYNC_EXAMPLE = AbortSignal.abort('aborted!');

describe('AbortSignal', () => {
  describe('serialize', () => {
    it('supports aborted AbortSignal', () => {
      const result = serialize(SYNC_EXAMPLE);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof SYNC_EXAMPLE>(result);
      expect(back).toBeInstanceOf(AbortSignal);
      expect(back.reason).toBe(SYNC_EXAMPLE.reason);
    });
    it('supports future AbortSignal', () => {
      const instance = delayedAbortSignal();
      const result = serialize(instance);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof instance>(result);
      expect(back).toBeInstanceOf(AbortSignal);
    });
  });
  describe('serializeAsync', () => {
    it('supports aborted AbortSignal', async () => {
      const result = await serializeAsync(Promise.resolve(SYNC_EXAMPLE));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof SYNC_EXAMPLE>>(result);
      expect(back).toBeInstanceOf(AbortSignal);
      expect(back.reason).toBe(SYNC_EXAMPLE.reason);
    });
    it('supports future AbortSignal', async () => {
      const instance = delayedAbortSignal();
      const result = await serializeAsync(instance);
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof instance>>(result);
      expect(back).toBeInstanceOf(AbortSignal);
    });
  });
  describe('toJSON', () => {
    it('supports aborted AbortSignal', () => {
      const result = toJSON(SYNC_EXAMPLE);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof SYNC_EXAMPLE>(result);
      expect(back).toBeInstanceOf(AbortSignal);
      expect(back.reason).toBe(SYNC_EXAMPLE.reason);
    });
    it('supports future AbortSignal', () => {
      const instance = delayedAbortSignal();
      const result = toJSON(instance);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof instance>(result);
      expect(back).toBeInstanceOf(AbortSignal);
    });
  });
  describe('toJSONAsync', () => {
    it('supports aborted AbortSignal', async () => {
      const result = await toJSONAsync(Promise.resolve(SYNC_EXAMPLE));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<typeof SYNC_EXAMPLE>>(result);
      expect(back).toBeInstanceOf(AbortSignal);
      expect(back.reason).toBe(SYNC_EXAMPLE.reason);
    });
    it('supports future AbortSignal', async () => {
      const instance = delayedAbortSignal();
      const result = await toJSONAsync(instance);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof instance>(result);
      expect(back).toBeInstanceOf(AbortSignal);
    });
  });

  describe('crossSerialize', () => {
    it('supports aborted AbortSignal', () => {
      const result = crossSerialize(SYNC_EXAMPLE);
      expect(result).toMatchSnapshot();
    });
    it('supports future AbortSignal', () => {
      const instance = delayedAbortSignal();
      const result = crossSerialize(instance);
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports aborted AbortSignal', () => {
        const result = crossSerialize(SYNC_EXAMPLE, { scopeId: 'example' });
        expect(result).toMatchSnapshot();
      });
      it('supports future AbortSignal', () => {
        const instance = delayedAbortSignal();
        const result = crossSerialize(instance, { scopeId: 'example' });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports aborted AbortSignal', async () => {
      const result = await crossSerializeAsync(Promise.resolve(SYNC_EXAMPLE));
      expect(result).toMatchSnapshot();
    });
    it('supports future AbortSignal', async () => {
      const instance = delayedAbortSignal();
      const result = await crossSerializeAsync(instance);
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports aborted AbortSignal', async () => {
        const result = await crossSerializeAsync(
          Promise.resolve(SYNC_EXAMPLE),
          {
            scopeId: 'example',
          },
        );
        expect(result).toMatchSnapshot();
      });
      it('supports future AbortSignal', async () => {
        const instance = delayedAbortSignal();
        const result = await crossSerializeAsync(instance, {
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports aborted AbortSignal', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(SYNC_EXAMPLE), {
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    it('supports future AbortSignal', async () =>
      new Promise<void>((resolve, reject) => {
        const instance = delayedAbortSignal();
        crossSerializeStream(instance, {
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports aborted AbortSignal', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(Promise.resolve(SYNC_EXAMPLE), {
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
      it('supports future AbortSignal', async () =>
        new Promise<void>((resolve, reject) => {
          const instance = delayedAbortSignal();
          crossSerializeStream(instance, {
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports aborted AbortSignal', () => {
      const result = toCrossJSON(SYNC_EXAMPLE);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof SYNC_EXAMPLE>(result, {
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(AbortSignal);
      expect(back.reason).toBe(SYNC_EXAMPLE.reason);
    });
    it('supports future AbortSignal', () => {
      const instance = delayedAbortSignal();
      const result = toCrossJSON(instance);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof instance>(result, {
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(AbortSignal);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports aborted AbortSignal', async () => {
      const result = await toCrossJSONAsync(Promise.resolve(SYNC_EXAMPLE));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromCrossJSON<Promise<typeof SYNC_EXAMPLE>>(result, {
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(AbortSignal);
      expect(back.reason).toBe(SYNC_EXAMPLE.reason);
    });
    it('supports future AbortSignal', async () => {
      const instance = delayedAbortSignal();
      const result = await toCrossJSONAsync(instance);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof SYNC_EXAMPLE>(result, {
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(AbortSignal);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports aborted AbortSignal', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(Promise.resolve(SYNC_EXAMPLE), {
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    it('supports future AbortSignal', async () =>
      new Promise<void>((resolve, reject) => {
        const instance = delayedAbortSignal();
        toCrossJSONStream(instance, {
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
