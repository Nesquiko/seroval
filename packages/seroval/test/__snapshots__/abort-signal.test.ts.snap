// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AbortSignal > crossSerialize > scoped > supports aborted AbortSignal 1`] = `"($R=>$R[0]=AbortSignal.abort("aborted!"))($R["example"])"`;

exports[`AbortSignal > crossSerialize > scoped > supports future AbortSignal 1`] = `"($R=>$R[0]=($R[1]=(a,s)=>((s=(a=new AbortController).signal).a=a,s))())($R["example"])"`;

exports[`AbortSignal > crossSerialize > supports aborted AbortSignal 1`] = `"$R[0]=AbortSignal.abort("aborted!")"`;

exports[`AbortSignal > crossSerialize > supports future AbortSignal 1`] = `"$R[0]=($R[1]=(a,s)=>((s=(a=new AbortController).signal).a=a,s))()"`;

exports[`AbortSignal > crossSerializeAsync > scoped > supports aborted AbortSignal 1`] = `"($R=>$R[0]=Promise.resolve($R[1]=AbortSignal.abort("aborted!")))($R["example"])"`;

exports[`AbortSignal > crossSerializeAsync > scoped > supports future AbortSignal 1`] = `"($R=>$R[0]=AbortSignal.abort($R[1]=Object.assign(new Error("The operation was aborted due to timeout"),{name:"TimeoutError",stack:"TimeoutError: The operation was aborted due to timeout\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\n    at listOnTimeout (node:internal/timers:573:17)\\n    at processTimers (node:internal/timers:514:7)"})))($R["example"])"`;

exports[`AbortSignal > crossSerializeAsync > supports aborted AbortSignal 1`] = `"$R[0]=Promise.resolve($R[1]=AbortSignal.abort("aborted!"))"`;

exports[`AbortSignal > crossSerializeAsync > supports future AbortSignal 1`] = `"$R[0]=AbortSignal.abort($R[1]=Object.assign(new Error("The operation was aborted due to timeout"),{name:"TimeoutError",stack:"TimeoutError: The operation was aborted due to timeout\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\n    at listOnTimeout (node:internal/timers:573:17)\\n    at processTimers (node:internal/timers:514:7)"}))"`;

exports[`AbortSignal > crossSerializeStream > scoped > supports aborted AbortSignal 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R["example"])"`;

exports[`AbortSignal > crossSerializeStream > scoped > supports aborted AbortSignal 2`] = `"($R=>($R[3]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=AbortSignal.abort("aborted!")))($R["example"])"`;

exports[`AbortSignal > crossSerializeStream > scoped > supports future AbortSignal 1`] = `"($R=>$R[0]=($R[1]=(a,s)=>((s=(a=new AbortController).signal).a=a,s))())($R["example"])"`;

exports[`AbortSignal > crossSerializeStream > scoped > supports future AbortSignal 2`] = `"($R=>($R[3]=(s,r)=>{s.a.abort(r);delete s.a})($R[0],$R[2]=Object.assign(new Error("The operation was aborted due to timeout"),{name:"TimeoutError",stack:"TimeoutError: The operation was aborted due to timeout\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\n    at listOnTimeout (node:internal/timers:573:17)\\n    at processTimers (node:internal/timers:514:7)"})))($R["example"])"`;

exports[`AbortSignal > crossSerializeStream > supports aborted AbortSignal 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`AbortSignal > crossSerializeStream > supports aborted AbortSignal 2`] = `"($R[3]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=AbortSignal.abort("aborted!"))"`;

exports[`AbortSignal > crossSerializeStream > supports future AbortSignal 1`] = `"$R[0]=($R[1]=(a,s)=>((s=(a=new AbortController).signal).a=a,s))()"`;

exports[`AbortSignal > crossSerializeStream > supports future AbortSignal 2`] = `"($R[3]=(s,r)=>{s.a.abort(r);delete s.a})($R[0],$R[2]=Object.assign(new Error("The operation was aborted due to timeout"),{name:"TimeoutError",stack:"TimeoutError: The operation was aborted due to timeout\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\n    at listOnTimeout (node:internal/timers:573:17)\\n    at processTimers (node:internal/timers:514:7)"}))"`;

exports[`AbortSignal > serialize > supports aborted AbortSignal 1`] = `"AbortSignal.abort("aborted!")"`;

exports[`AbortSignal > serialize > supports future AbortSignal 1`] = `"(h=>((a,s)=>((s=(a=new AbortController).signal).a=a,s))())()"`;

exports[`AbortSignal > serializeAsync > supports aborted AbortSignal 1`] = `"Promise.resolve(AbortSignal.abort("aborted!"))"`;

exports[`AbortSignal > serializeAsync > supports future AbortSignal 1`] = `"AbortSignal.abort(Object.assign(new Error("The operation was aborted due to timeout"),{name:"TimeoutError",stack:"TimeoutError: The operation was aborted due to timeout\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\n    at listOnTimeout (node:internal/timers:573:17)\\n    at processTimers (node:internal/timers:514:7)"}))"`;

exports[`AbortSignal > toCrossJSON > supports aborted AbortSignal 1`] = `"{"t":37,"i":0,"f":{"t":1,"s":"aborted!"}}"`;

exports[`AbortSignal > toCrossJSON > supports future AbortSignal 1`] = `"{"t":35,"i":0,"f":{"t":26,"i":1,"s":5}}"`;

exports[`AbortSignal > toCrossJSONAsync > supports aborted AbortSignal 1`] = `"{"t":12,"i":0,"s":1,"f":{"t":37,"i":1,"f":{"t":1,"s":"aborted!"}}}"`;

exports[`AbortSignal > toCrossJSONAsync > supports future AbortSignal 1`] = `"{"t":37,"i":0,"f":{"t":13,"i":1,"s":0,"m":"The operation was aborted due to timeout","p":{"k":["name","stack"],"v":[{"t":1,"s":"TimeoutError"},{"t":1,"s":"TimeoutError: The operation was aborted due to timeout\\\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\\\n    at listOnTimeout (node:internal/timers:573:17)\\\\n    at processTimers (node:internal/timers:514:7)"}],"s":2}}}"`;

exports[`AbortSignal > toCrossJSONStream > supports aborted AbortSignal 1`] = `"{"t":22,"i":0,"f":{"t":26,"i":1,"s":1}}"`;

exports[`AbortSignal > toCrossJSONStream > supports aborted AbortSignal 2`] = `"{"t":23,"i":0,"a":[{"t":26,"i":3,"s":2},{"t":37,"i":2,"f":{"t":1,"s":"aborted!"}}]}"`;

exports[`AbortSignal > toCrossJSONStream > supports future AbortSignal 1`] = `"{"t":35,"i":0,"f":{"t":26,"i":1,"s":5}}"`;

exports[`AbortSignal > toCrossJSONStream > supports future AbortSignal 2`] = `"{"t":36,"i":0,"a":[{"t":26,"i":3,"s":6},{"t":13,"i":2,"s":0,"m":"The operation was aborted due to timeout","p":{"k":["name","stack"],"v":[{"t":1,"s":"TimeoutError"},{"t":1,"s":"TimeoutError: The operation was aborted due to timeout\\\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\\\n    at listOnTimeout (node:internal/timers:573:17)\\\\n    at processTimers (node:internal/timers:514:7)"}],"s":2}}]}"`;

exports[`AbortSignal > toJSON > supports aborted AbortSignal 1`] = `"{"t":{"t":37,"i":0,"f":{"t":1,"s":"aborted!"}},"f":47,"m":[]}"`;

exports[`AbortSignal > toJSON > supports future AbortSignal 1`] = `"{"t":{"t":35,"i":0,"f":{"t":26,"i":1,"s":5}},"f":47,"m":[]}"`;

exports[`AbortSignal > toJSONAsync > supports aborted AbortSignal 1`] = `"{"t":{"t":12,"i":0,"s":1,"f":{"t":37,"i":1,"f":{"t":1,"s":"aborted!"}}},"f":47,"m":[]}"`;

exports[`AbortSignal > toJSONAsync > supports future AbortSignal 1`] = `"{"t":{"t":37,"i":0,"f":{"t":13,"i":1,"s":0,"m":"The operation was aborted due to timeout","p":{"k":["name","stack"],"v":[{"t":1,"s":"TimeoutError"},{"t":1,"s":"TimeoutError: The operation was aborted due to timeout\\\\n    at new DOMException (node:internal/per_context/domexception:53:5)\\\\n    at Timeout._onTimeout (node:internal/abort_controller:130:9)\\\\n    at listOnTimeout (node:internal/timers:573:17)\\\\n    at processTimers (node:internal/timers:514:7)"}],"s":2}}},"f":47,"m":[]}"`;
